### ISSUE TITLE
Nested factors

### ISSUE THREAD

**Issue by BenAbabaei** [APSIMInitiative/ApsimX#3378]
I have a large set of simulations to run and I need to efficiently use factors. I am interested in creating nested factors. In the following example, I have three factors; **Site**, **ISM** (initial soil moisture) and **Sowing** (sowing date and cultivar). In each location, sowing date and agronomic management are different. When I put **Sowing** under the **Management** folder, everything works fine. But I need to use multiple **genotypes** and that's why I want to have **Sowing** as a factor (Sowing = combination of sowing date and cultivar). 

- **Is there any way to have all the combinations of factors in each site, but different combinations in different sites?** 
- **Is there any way to link a property in a manager script (e.g. cultivar name) to a factor**? 
- **Is there any way to use Sowing Date and Cultivar as two separate factors**? 

![image](https://user-images.githubusercontent.com/39573622/49721938-3268b780-fcaf-11e8-9c16-268c1364a578.png)

**Comment by BenAbabaei** [comment]
The following settings seem working. But I don't understand how it is that the interface treats the two last objects (i.e. **Cultivar1** and **Cultivar2**) as a factor? I just put **[Sowing]** on one line inside the factor **Tehran**!

![image](https://user-images.githubusercontent.com/39573622/49722404-64c6e480-fcb0-11e8-9a13-68cfa11e36e9.png)

![image](https://user-images.githubusercontent.com/39573622/49722497-a9528000-fcb0-11e8-8879-5f21bfe1d68e.png)

**Comment by hol353** [comment]
Is this with a recent version i.e. last 3 weeks?

**Comment by hol353** [comment]
Can you upload your .apsimx file to this issue?

**Comment by BenAbabaei** [comment]
@hol353 the sample you asked for.
[Temp.zip](https://github.com/APSIMInitiative/ApsimX/files/2665219/Temp.zip)

**Comment by hol430** [comment]
> Is there any way to have all the combinations of factors in each site, but different combinations in different sites?

Unsure exactly what you mean here. As an example, you should be able to change the weather in one site, and change a soil in another site. That *should* work...is that what you mean?

> Is there any way to link a property in a manager script (e.g. cultivar name) to a factor?

I think so. Have a look at the `AlternativeSowing` factor in the attached file.

> Is there any way to use Sowing Date and Cultivar as two separate factors?

Pretty sure this should be possible. Again, look at `AlternativeSowing` in the attached file. It will modify the cultivar. Then just copy that factor and change the sowing date in the copy.


Hopefully that makes sense....let me know if you run into any problems.

Also I deleted the hourly tr function from the attached file so that I could open it in my branch...you'll presumably need to add it back in.

[TempSanitised.zip](https://github.com/APSIMInitiative/ApsimX/files/2794758/TempSanitised.zip)

**Comment by BenAbabaei** [comment]
Thanks Drew. That **AlternativeSowing** answers a couple of questions I had. As to the first point, I needed to have specific combinations of factors. For example, I had 60 sites, each with their own weather data and soil. So, I needed to use 60 soils and 60 met files without having 60*60 combinations of them. I finally put them in a folder. That was the only way I could do it. Cheers.

![image](https://user-images.githubusercontent.com/39573622/51727562-48440780-20b8-11e9-8314-97942ecf3b2d.png)

**Comment by hol430** [comment]
So you want these manager scripts in one factor, but not in other factors or the base simulation? The soil and weather nodes seem to work fine, and are not in a folder.

I also notice that you have [Sowing] in the Tehran factor, even though none of the models inside the factor replace the Sowing script. This seems to be causing some subtle problems

**Comment by BenAbabaei** [comment]
I am not sure which version of file you are referring to. What we needed to do was to simulate 60 locations in each of which sowing, fertilisation and weather was different. The only way I could do that without having all the combinations of 60 weathers x 60 soils x 60 managements was to use **folders**. I put them in one folder and linked that folder to an empty folder in the base simulation.

**Comment by hol430** [comment]
Ah, I see the problem here - we are trying to use factors in two different ways:

1. I want to replace one model with multiple replacements
![image](https://user-images.githubusercontent.com/36427516/52311589-338a3c80-29f3-11e9-9086-18c94ab3b017.png)

2. I want to replace multiple models, each with a single replacement
![image](https://user-images.githubusercontent.com/36427516/52311609-469d0c80-29f3-11e9-8610-b8a967020e9d.png)

In scenario 1, Apsim goes looking for all manager scripts under this factor, and each of these manager scripts will be used as a replacement for ResetWaterOnSowing. 

In scenario 2, we want to provide a ***single*** replacement for the sowing manager script, and a ***single*** replacement for the FertiliserZ37 manager script. However, the same logic applies - Apsim goes looking for all manager scripts under this factor, and each of these manager scripts will be used to replace Sowing, and each of these manager scripts will also be used as a replacement for FertiliserZ37, so you end up with something like this:

![image](https://user-images.githubusercontent.com/36427516/52310833-81517580-29f0-11e9-8ba9-48e5add4bd1b.png)

What I propose is that if the factor node holds only one factor specification (that is, your factor node only contains one line, as in the case of scenario 1), we perform multiple replacements for the single specification. Otherwise, if the factor node contains multiple specifications, we perform a single replacement for each specification/child node. This appears to fix your problem:

![image](https://user-images.githubusercontent.com/36427516/52311796-f70b1080-29f3-11e9-835f-4c9fc632ff47.png)


The only catch is that in scenario 2, each child model of the factor will have to have the correct name. e.g. Sowing1 will not work, because the model you are trying to replace is called Sowing. 

@hol353 what do you think?

**Comment by BenAbabaei** [comment]
Dear Drew. You are right (**G! I had written WRITE at first!**). That may work. This is how I did it a couple of weeks ago. I had multiple locations in which I needed to specify Soil, Weather, and Management (i.e. multiple manager scripts). I had to put manager scripts in a folder to make sure they are not combined with other factors leading to many unwanted simulations. I really cannot remember what would have happened if I had put those scripts out of the folder. I just remember it was not possible!

[TEHartog-2.zip](https://github.com/APSIMInitiative/ApsimX/files/2834330/TEHartog-2.zip)

![image](https://user-images.githubusercontent.com/39573622/52312523-93361700-29f6-11e9-8efe-d41f04a4a8b9.png)
